scp/lib/bfs/bucket/scp.rb

```diff
@@ -145,32 +145,53 @@ def mkdir_p(path)
       end
 
       def sh!(*cmd) # rubocop:disable Metrics/MethodLength
-        stdout = ''
         stderr = nil
         status = 0
         cmdstr = cmd.map {|x| Shellwords.escape(x) }.join(' ')
 
+        # init command's STDOUT stream (large output may break lines in on_data):
+        stdout_reader, stdout_writer = IO.pipe
```

Comment:
I couldn't find anything more proper (in-memory stream piping like Go's https://golang.org/pkg/io/#Pipe) - Ruby's IO.pipe looks to be backed by OS pipes:

https://ruby-doc.org/core-3.0.0/IO.html#method-c-pipe 

> Not available on all platforms.

Heavy or not, it should do the trick. Maybe we could swap it with some custom-written in-mem pipe (haven't found anything like that for Ruby) if that ever becomes a concern.

Reply:
I don't think you need to do it, you can do something much simpler and accept a block to allow chunking:

```ruby
def sh!(*cmd)
...
   ch.on_data do |_, data|
      yield data
   end
```

basically, instead of concatting into stdout and returning it, let the caller process chunks of stdout instead

Reply:
@dim that was the original approach, and lines were broken for large output. Hence streaming (stream can combine 2 chunks with broken line in the middle, so each_line works as expected)

Reply:
k, checked again, I still don't like the abuse of pipe that is happening here. it doesn't need to be that complicated, you can just do something like this:

```rb
ch.on_data do |_, data|
  stdout << data
  if block_given?
    pos = stdout.index("\n")    
    yield stdout.slice!(..pos) if pos
  end
end
```

Reply:
:man_facepalming: right